### PR TITLE
Refactor implementation of colocation feature.

### PR DESCRIPTION
**Pull request by rwjblue**
* Extracts helper method (to simplify processing of component files with templates)
* Add guard for when colocation should not be enabled (still somewhat WIP, but the branching is setup).
* Change the implementation of colocation to rely on `hbs` (still WIP)

---

Note: This targets the `colocation` branch (so that we can avoid breaking folks actively using the `@ember/octane-app-blueprint` which uses the `colocation` branch).

### CODE REVIEW

File: colocated-broccoli-plugin.js [hunk 1]
```diff
@@ -14,6 +15,21 @@ module.exports = class ColocatedTemplateProcessor extends Plugin {
...
+    if (!jsContents.includes('export default')) {
+      let message = `\`${filePath}\` does not contain a \`default export\`. Did you forget to export the component class?`;
+      return `${jsContents}\nthrow new Error(${JSON.stringify(message)});`;
+    }
+
+    // TODO: deal with hygiene?
+    let output = stripIndents`
+      import hbs from 'ember-cli-htmlbars-inline-precompile';
+      const __COLOCATED_TEMPLATE__ = hbs\`${templateContents}\`;
+      ${jsContents}`;
```
Comment (chancancode) [inline colocated-broccoli-plugin.js]: Does `stripIndents` look at the template string, or does it look at the interpolated string? (i.e. does the value/content of what is inside of `jsContents` affect the stripping?)

File: colocated-broccoli-plugin.js [hunk 2]
```diff
@@ -14,6 +15,21 @@ module.exports = class ColocatedTemplateProcessor extends Plugin {
     this.options = options;
   }
 
+  _transformExistingComponent(filePath, jsContents, templateContents) {
+    if (!jsContents.includes('export default')) {
+      let message = `\`${filePath}\` does not contain a \`default export\`. Did you forget to export the component class?`;
+      return `${jsContents}\nthrow new Error(${JSON.stringify(message)});`;
+    }
```
Comment (chancancode) [inline colocated-broccoli-plugin.js]: We can probably move this check/rewrite to the babel plugin now, for fewer false positives.

File: colocated-broccoli-plugin.js [hunk 3]
```diff
@@ -52,30 +68,21 @@ module.exports = class ColocatedTemplateProcessor extends Plugin {
...
-const TEMPLATE = ${this.options.precompile(templateContents)};
-const CLASS = templateOnlyComponent();
-export default setComponentTemplate(TEMPLATE, CLASS);`;
+
+        jsContents = stripIndents`
+          import templateOnly from '@ember/component/template-only';
+          import hbs from 'ember-cli-htmlbars-inline-precompile';
+          const __COLOCATED_TEMPLATE__ = hbs\`${templateContents}\`;
+
+          export default templateOnly();`
```
Comment (chancancode) [inline colocated-broccoli-plugin.js]: training new line

File: colocated-broccoli-plugin.js (hunk 1 above)
Comment (chancancode) [reply colocated-broccoli-plugin.js]: (this one does not need trailing newline since `jsContents` should already contain one)

Comment (rwjblue) [reply colocated-broccoli-plugin.js]: @chancancode - It uses the full interpolated value _before_ stripping all leading indentation.

Comment (rwjblue) [reply colocated-broccoli-plugin.js]: I'll add tests confirming that we don't muck up all the leading whitespace in the `jsContents` though. Possibly changing this to:

```js
    let output = stripIndents`
      import hbs from 'ember-cli-htmlbars-inline-precompile';
      const __COLOCATED_TEMPLATE__ = hbs\`${templateContents}\`;
      ` + jsContents;
```

File: colocated-broccoli-plugin.js (hunk 2 above)
Comment (rwjblue) [reply colocated-broccoli-plugin.js]: Good idea!

File: colocated-broccoli-plugin.js (hunk 1 above)
Comment (chancancode) [reply colocated-broccoli-plugin.js]: Yeah in that case I think (?) the current code is not correct, since jsContents likely have lines without leading indentation, I think it would cause the whole thing to not get stripped properly.

Comment (rwjblue) [reply colocated-broccoli-plugin.js]: Yep, confirm. I’ll add a test (that will fail), then fix the impl to my suggested thing above.

**Comment by rwjblue** [comment]
FYI - This is a prototype of the AST transform: 

https://astexplorer.net/#/gist/bcca584efdab6c981a75618642c76a22/8fd5932d3853afa959f902e1e25b4526665af18d

File: colocated-broccoli-plugin.js (hunk 1 above)
Comment (rwjblue) [reply colocated-broccoli-plugin.js]: This is fixed (no longer using any indentation stripping).
